File: AjaxVectorObjects/ClientScripts/Math/PointF.js
// Copyright (c) 2018 Aurigma Inc. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
//
Type.registerNamespace("Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math");

Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF = function (x, y) {
    ///	<summary>This client-side class corresponds to the <see cref="T:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF" /> server-side class.</summary>
    /// <remarks><para>This class represents an ordered pair of floating-point x- and y-coordinates that defines a point in a two-dimensional plane and exposes a number of methods to operate with it.</para></remarks>
    /// <param name="x" type="Number" />
    /// <param name="y" type="Number" />
    /// <field name="X" type="Number" integer="true" static="true"><summary>The x-coordinate of this point.</summary></field>
    /// <field name="Y" type="Number" integer="true" static="true"><summary>The y-coordinate of this point.</summary></field>
    /// <seealso cref="T:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF" />
    /// <constructor>
    /// 	<summary>Creates and initializes an instance of the <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF" /> class.</summary>
    /// 	<param name="x">The x-coordinate of the point to create.</param>
    /// 	<param name="y">The y-coordinate of the point to create.</param>
    /// </constructor>
    this.X = (x) ? x : 0;
    this.Y = (y) ? y : 0;
};

Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF.prototype = {
    rotate: function (angle) {
        /// <summary>Rotates a point to the specified angle.</summary>
        /// <param name="angle" type="Number">The angle to rotate point to.</param>
        /// <returns type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF">The <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF" /> value which represents a rotated point.</returns>
        /// <remarks><para>This method corresponds to <see cref="M:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF.Rotate(System.Single)">PointF.Rotate(Single)</see> server-side member.</para></remarks>
        //        var radianAngle = Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.ConvertDegreeToRadian(angle);
        //        var newX = this.X * Math.cos(radianAngle) - this.Y * Math.sin(radianAngle);
        //        var newY = this.X * Math.sin(radianAngle) + this.Y * Math.cos(radianAngle);
        //        this.X = newX;
        //        this.Y = newY;
        return this.rotateAt(angle, { X: 0, Y: 0 });
    },

    rotateAt: function (angle, center) {
        if (!center)
            center = { X: 0, Y: 0 };
        var radianAngle = Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.ConvertDegreeToRadian(angle);
        var newX = Math.cos(radianAngle) * (this.X - center.X) - Math.sin(radianAngle) * (this.Y - center.Y) + center.X;
        var newY = Math.sin(radianAngle) * (this.X - center.X) + Math.cos(radianAngle) * (this.Y - center.Y) + center.Y;
        this.X = newX;
        this.Y = newY;
        return this;
    },

    translate: function (x, y) {
        /// <summary>Translates a point to the specified offset.</summary>
        /// <param name="x" type="Number">The offset in the x-direction.</param>
        /// <param name="y" type="Number">The offset in the y-direction.</param>
        /// <returns type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF">The <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF" /> value which represents a translated point.</returns>
        /// <remarks><para>This method corresponds to <see cref="M:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF.Translate(System.Single,System.Single)">PointF.Translate(Single, Single)</see> server-side member.</para></remarks>
        this.X = this.X + x;
        this.Y = this.Y + y;
        return this;
    },

    scale: function (scaleX, scaleY) {
        /// <summary>Scales a point to the specified values.</summary>
        /// <param name="scaleX" type="Number">The scale factor in the x-direction.</param>
        /// <param name="scaleY" type="Number">The scale factor in the y-direction.</param>
        /// <returns type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF">The <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF" /> value which represents a scaled point.</returns>
        /// <remarks><para>This method corresponds to <see cref="M:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF.Scale(System.Single,System.Single)">PointF.Scale(Single, Single)</see> server-side member.</para></remarks>
        this.X = this.X * scaleX;
        this.Y = this.Y * scaleY;
        return this;
    },

    clone: function () {
        /// <summary>Creates a new object that is a copy of the current <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF" /> instance.</summary>
        /// <returns type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF">A new <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF" /> that is a copy of this instance.</returns>
        return new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF(this.X, this.Y);
    },

    isEqual: function (pt, tolerance) {
        /// <summary>Determines whether the specified <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF" /> is equal to the current <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF" />.</summary>
        /// <param name="pt" type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF">The point to compare with the current <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF" />.</param>
        /// <returns type="Boolean"><strong>true</strong> if <paramref name="pt" /> is equal to the current <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF" />; otherwise, <strong>false</strong>.</returns>

        var m = Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math;
        return pt != null && m.EqualsOfFloatNumbers(this.X, pt.X, tolerance) && m.EqualsOfFloatNumbers(this.Y, pt.Y, tolerance);
    },

    distance: function (pt) {
        /// <summary>Returns a distance from this point to the specified one.</summary>
        /// <param name="pt" type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF">The destination point.</param>
        /// <returns type="Number">The number which represents a distance between this point and <paramref name="point" />.</returns>
        /// <remarks><para>This method corresponds to <see cref="M:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF.Distance(Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF)">PointF.Distance(PointF)</see> server-side member.</para></remarks>
        return Math.sqrt((this.X - pt.X) * (this.X - pt.X) + (this.Y - pt.Y) * (this.Y - pt.Y));
    },

    transform: function (transform, center) {
        if (center == null)
            center = new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF();

        this.translate(-center.X, -center.Y);
        this.scale(transform.get_scaleX(), transform.get_scaleY());
        this.rotate(transform.get_angle());
        this.translate(transform.get_translateX(), transform.get_translateY());
        this.translate(center.X, center.Y);

        return this;
    },

    toString: function () {
        var p = [this.X, this.Y];
        return p.join(',');
    },

    round: function () {
        this.X = Math.round(this.X);
        this.Y = Math.round(this.Y);
        return this;
    },

    X: 0,
    Y: 0
}

Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF.registerClass("Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math.PointF");

File: AjaxVectorObjects/Math/PointF.cs
// Copyright (c) 2018 Aurigma Inc. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
//
using System;
using System.Drawing;
using System.Globalization;

namespace Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math
{
    public class PointF : ICloneable
    {
        public PointF()
        {
        }

        public PointF(float x, float y)
        {
            X = x;
            Y = y;
        }

        public PointF(PointF pt)
        {
            X = pt.X;
            Y = pt.Y;
        }

        public PointF(System.Drawing.PointF pt)
        {
            X = pt.X;
            Y = pt.Y;
        }

        public float X { get; set; }

        public float Y { get; set; }

        /// <summary>
        /// Rotate point
        /// </summary>
        /// <param name="angle">Angle in degrees</param>
        /// <returns>Point itself</returns>
        public PointF Rotate(float angle)
        {
            return Rotate((double)angle);
        }

        internal PointF Rotate(double angle)
        {
            return RotateAt(angle, new PointF(0, 0));
        }

        internal PointF RotateAt(double angle, PointF center)
        {
            angle = Utils.ConvertDegreeToRadian(angle);
            var pt = new PointF(X, Y);
            X = (float)(System.Math.Cos(angle) * (pt.X - center.X) - System.Math.Sin(angle) * (pt.Y - center.Y) + center.X);
            Y = (float)(System.Math.Sin(angle) * (pt.X - center.X) + System.Math.Cos(angle) * (pt.Y - center.Y) + center.Y);
            return this;
        }

        public PointF Translate(float x, float y)
        {
            return Translate(x, (double)y);
        }

        internal PointF Translate(double x, double y)
        {
            X = (float)(x + X);
            Y = (float)(y + Y);
            return this;
        }

        public PointF Scale(float scaleX, float scaleY)
        {
            return Scale(scaleX, (double)scaleY);
        }

        internal PointF Scale(double scaleX, double scaleY)
        {
            X = (float)(scaleX * X);
            Y = (float)(scaleY * Y);
            return this;
        }

        internal PointF Transform(Transform transform, PointF center)
        {
            Translate(-center.X, -center.Y);
            Scale(transform.ScaleX, transform.ScaleY);
            Rotate(transform.Angle);
            Translate(transform.TranslateX, transform.TranslateY);
            Translate(center.X, center.Y);

            return this;
        }

        public double Distance(PointF point)
        {
            return System.Math.Sqrt((X - point.X) * (X - point.X) + (Y - point.Y) * (Y - point.Y));
        }

        public Point ToPoint()
        {
            return new Point(Convert.ToInt32(X), Convert.ToInt32(Y));
        }

        public System.Drawing.PointF ToPointF()
        {
            return new System.Drawing.PointF(X, Y);
        }

        public PointF Clone()
        {
            return new PointF(this);
        }

        public static bool AreEqualValues(PointF p1, PointF p2, double tolerance = 0.0001)
        {
            if (ReferenceEquals(p1, p2))
            {
                return true;
            }

            if (!ReferenceEquals(p1, null) && !ReferenceEquals(p2, null))
            {
                return Utils.EqualsOfFloatNumbers(p1.X, p2.X, tolerance) && Utils.EqualsOfFloatNumbers(p1.Y, p2.Y, tolerance);
            }

            return false;
        }

        public bool IsEmpty
        {
            get
            {
                return (Utils.EqualsOfFloatNumbers(X, 0) && Utils.EqualsOfFloatNumbers(Y, 0));
            }
        }

        public bool Equals(PointF p, double tolerance = 0.0001)
        {
            return AreEqualValues(this, p, tolerance);
        }

        public override string ToString()
        {
            return string.Format(CultureInfo.InvariantCulture, "{{X={0}, Y={1}}}", X, Y);
        }

        #region ICloneable Members

        object ICloneable.Clone()
        {
            var p = Clone();
            return p;
        }

        #endregion ICloneable Members
    }
}